java-checks/src/main/java/org/sonar/java/checks/ThisExposedFromConstructorCheck.java

```diff
@@ -0,0 +1,104 @@
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import org.sonar.check.Rule;
+import org.sonar.java.model.ExpressionUtils;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S3366")
+public class ThisExposedFromConstructorCheck extends IssuableSubscriptionVisitor {
+
+  private static final String REPORT = "Make sure the use of \"this\" doesn't" +
+    " expose partially-constructed instances of this class in multi-threaded environments.";
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.CONSTRUCTOR);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    MethodTree methodTree = (MethodTree) tree;
+    methodTree.block().accept(new ConstructorBodyVisitor(methodTree.symbol().owner()));
+  }
+
+  private class ConstructorBodyVisitor extends BaseTreeVisitor {
+    private Symbol owner;
+
+    public ConstructorBodyVisitor(Symbol owner) {
+      this.owner = owner;
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree tree) {
+      if (this.owner == tree.symbol().owner()) {
+        return;
+      }
+      tree.arguments().stream().filter(this::isThis).forEach(expr -> reportIssue(expr, REPORT));
+    }
+
+    @Override
+    public void visitAssignmentExpression(AssignmentExpressionTree tree) {
+      if (!isThis(tree.expression())) {
+        return;
+      }
+      ExpressionTree variable = tree.variable();
+      if (variable.is(Tree.Kind.MEMBER_SELECT)) {
+        MemberSelectExpressionTree memberSelect = (MemberSelectExpressionTree) variable;
+        if (memberSelect.expression().symbolType().symbol().equals(this.owner) && memberSelect.identifier().symbol().isStatic()) {
+          return;
+        }
+      } else if (variable.is(Tree.Kind.IDENTIFIER) && (((IdentifierTree) variable).symbol().isStatic())) {
+        return;
+      }
+      reportIssue(tree, REPORT);
+      super.visitAssignmentExpression(tree);
+    }
+
+    @Override
+    public void visitClass(ClassTree tree) {
+      // skip nested and anonymous classes
+    }
+
+    private boolean isThis(ExpressionTree expression) {
+      if (expression.is(Tree.Kind.IDENTIFIER)) {
+        String targetName = ((IdentifierTree) (ExpressionUtils.skipParentheses(expression))).name();
```

Comment:
Please add a case where the skipParentheses call is necessary. (in other words : with your current unit test, I can remove the call to ExpressionUtils.skipParentheses and the tests would not fail).